### ISSUE TITLE
Should not send all errors into the channel

### ISSUE THREAD

**Issue by matthijskooijman** [embolalia/willie#784]
Right now, if any module triggers an exception, it is sent back to the channel that triggered the error. This is useful for when you actually interact with the bot through commands, but for rules that are not intended to generate a response (like the chanlog module), this behaviour is unexpected and potentially spammy. As an example:

```
12:32 <TheSeven> A124: yes, those 20°C make a lot of difference for these big joints... I usually use 270-280 for SMD work and 300-320 for bigger stuff
12:32 <grobbebol> UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 51: ordinal not in range(128) (file "/data/applications/grobbebol/willie/willie/modules/chanlogs.py", line 135, in log_message)
```

This is really uncalled for. Perhaps these "silent" rules should somehow declare their silentness, so willie can refrain from spamming a channel with errors (and perhaps spam the owner instead)?

**Comment by embolalia** [comment]
I've addressed this [before](https://github.com/embolalia/willie/issues/717#issuecomment-75387124) (fourth paragraph). TL;DR, WONTFIX.

**Comment by flexd** [comment]
While I understand this has been addressed before and you won't fix it. Why can't the spewing of errors to publi channels be a toggleable option? In the other thread someone says you might miss exceptions. Isn't that why there is a exception log? I do not see understand why this cannot be something optional.

**Comment by michaelshmitty** [comment]
Sopel should not spit out exceptions to a channel. Just like any Linux program will output errors to STDERR and not STDOUT.
I have read #717 thoroughly and I agree with user MyWay that this *is* a security issue. You don't leak internal operating system information into a public channel, period.
Obviously this is something several Sopel users request. Closing #717 in the way it was done will not make this go away, as this newer issue indicates.

Nobody is demanding you to fix this, I'm sure all Sopel users are extremely grateful for this great piece of open source software, they're just asking you to consider this feature to be put on the roadmap so someone can pick it up.

If this has been fixed in the meantime, consider my comment void.

Thanks

**Comment by flexd** [comment]
Just to be clear: The fact that #717 was closed the way it is, and the unwillingness of the author to even add a configuration option to disable the error messages being posted to public channels is the reason I decided against using this library. They are obviously not going to listen, I would just find some other library. I ended up writing the bot I wanted in Go instead.

**Comment by dgw** [comment]
Apparently I didn't get to the other conversation on this, but here's how I see it.

Technical error messages, like exceptions, are a bad user experience for the vast majority of users in a channel. They're not really helpful to anyone but the bot maintainer, usually. If my bot throws an error while I'm offline, I inevitably get a `.tell` message when I come back with a copy-and-paste of the exception and something like "fix this pls, I don't know what it means".

It would probably be best if the code pointed out by @flexd simply output a general error message telling `trigger.nick` to tell the bot owner that something went wrong (possibly excluding `event`-based triggers, so things like join/part/quit/nick handlers don't spam people).

Exceptions are useful to two groups of people: bot admins and attackers. Bot admins can use them to find where an error is occurring, and either fix it or report it to the author of that code. Attackers can use them to figure out what OS the bot is running on, for example, or trick poorly coded modules that access the filesystem into possibly disclosing files they shouldn't. Long shot? Maybe, but a bot running under a user's account can read at least everything in that user's home directory, so it's worth protecting users from themselves IMO.

Bucket, #xkcd's (on Foonetic) channel bot, logs errors to a special logging channel. Doing so allows it to be restricted to trusted users if desired. I'm not suggesting that Sopel *require* a separate logging channel, but having the option would allow those of us who want an extra layer of security to get it while still getting the benefit of not needing to open a log file to see exceptions.

Would I be happy to contribute to this? Sure. But given the current attitude toward it by the project maintainers, I'm not going to work on any sort of pull request because at the moment it feels like such contributions would simply be rejected with a "no thanks".

**Comment by maxpowa** [comment]
FWIW, I agree with at least having the option to prevent the output of errors. I'm not going to go in depth as @dgw raised most of the points I would have but for anyone who's stumbling upon this and just wants to stop replying with the error message, check out [this diff](https://github.com/maxpowa/Inumuta/commit/3f524604dc2f3988968c6ef6ce4a49500c46c7d0). The diff linked simply redirects all errors to the configured `logging_channel`.

@embolalia Even though this has been raised in the past and you're not wrong, I think that at least having the ability to configure the behavior here is reasonable.

**Comment by thomastanck** [comment]
Just adding my vote in:

I feel that it's unprofessional to have error messages show up in chat, whether there is a security issue or not.

I also feel that new module developers that accidentally a security issue can and probably will be more exposed if error messages show up in chat. Yes, the security issue is elsewhere, but you don't broadcast the fact that your site is vulnerable to SQL injection for all of the public to see.

Finally, I feel that it's probably a simple fix, (indeed, @maxpowa is providing a 5 line diff right now!) and it's an utter waste of everyone's time to talk about fixing something this trivial.

Cough. https://en.wikipedia.org/wiki/Parkinson's_law_of_triviality

**Comment by dgw** [comment]
As a follow-up to this long-closed issue, I would like to note that the config setting `core.reply_errors` has been available since Sopel 6.4.0. It was introduced by #1071. Ultimately, this was resolved, not WONTFIX'd.
